doc: update and re-structure documenation (#325)
intel/meta-acrn: docs/getting-started.md

# Getting Started

> This guide will show how to setup host machine for Yocto build and then how to build and boot ACRN Hypervisor, Service VM and User VMs on Intel platforms.

Reviewer: "how to setup host machine" --> "how to set up the host machine" (the verb form is "set up"; the noun form is "setup")

   - [Configure Post-launched User VM (UOS)](#configure-post-launched-user-vm-uos)
   - [Build User VM (UOS) image via multiconfig](#build-uos-image-via-multiconfig)
   - [Build ACRN image](#build-acrn-image)

IV.  [Booting ACRN Image](#booting-acrn-image)
 * [Boot ACRN and SOS](#boot-acrn-and-sos)
 * [Launching UOS](#launching-uos)
   - [USB Passthrough](#usb-passthrough)

V.   [Configurations](#configurations)
 * [MACHINE Configuration](#machine-configuration)
 * [Adding Guests](#adding-guests)
 * [ACRN Configuration](#acrn-configuration)
   - [ACRN BOARD Configuration](#acrn-board-configuration)
   - [ACRN SCENARIO Configuration](#acrn-scenario-configuration)
   - [ACRN BUILD MODE Configuration](#acrn-build-mode-configuration)
 * [Kernel Configuration](#kernel-configuration)
   - [SOS](#sos)
   - [UOS](#uos)
 * [GRUB Configuration](#grub-configuration)
 * [Override distro configuration](#override-distro-configuration)

VI.  [Build WIC Installer Image](#build-wic-installer-image)

VII. [Tested Hardware](#tested-hardware)

VIII.[Contributing](#contributing)

## Overview
This layer provides ACRN Hypvervisor integration with Yocto Project.

## Supported Hardware

> ACRN is supported on below Intel platforms:

Reviewer: "on below" --> "on the following"

* Kaby Lake
* Whiskey Lake
* Tiger Lake
* Elkhart Lake

About minimum system requirements and limitations, please find more information at [Supported Hardware](https://projectacrn.github.io/latest/reference/hardware.html)

## Setup Build Host

> 

Reviewer: "Setup Build Host" --> "Set Up Build Host"

$ git clone https://git.yoctoproject.org/git/meta-virtualization
$ git clone https://git.yoctoproject.org/git/meta-intel
$ git clone https://github.com/intel/meta-acrn.git
```
* Initialize build environment
```
$ source oe-init-build-env
```

* Add layers to build environment `conf/bblayers.conf`
```
$ bitbake-layers add-layer ../meta-openembedded/meta-oe
$ bitbake-layers add-layer ../meta-openembedded/meta-python
$ bitbake-layers add-layer ../meta-openembedded/meta-filesystems
$ bitbake-layers add-layer ../meta-openembedded/meta-networking
$ bitbake-layers add-layer ../meta-virtualization
$ bitbake-layers add-layer ../meta-intel
$ bitbake-layers add-layer ../meta-acrn
```

#### Configure Service VM (SOS)

meta-acrn maintains prototype DISTRO configurations for both Servcie VM OS (SOS) and User VM OS (UOS). `acrn-demo-sos` for Service VM OS and `acrn-demo-uos` for User VM OS.
`local.conf` carries the common configuration, which can be overwritten by individual multiconfigs `conf/multiconfig/xxx.conf`
For the Service VM OS, we carry configuration in `conf/local.conf` and for User VM OS we carry in `conf/multiconfig/uos.conf`

> Append below configuration in `conf/local.conf`:

Reviewer: "Append below" --> "Append the following"

   - [Build ACRN image](#build-acrn-image)

IV.  [Booting ACRN Image](#booting-acrn-image)
 * [Boot ACRN and SOS](#boot-acrn-and-sos)
 * [Launching UOS](#launching-uos)
   - [USB Passthrough](#usb-passthrough)

V.   [Configurations](#configurations)
 * [MACHINE Configuration](#machine-configuration)
 * [Adding Guests](#adding-guests)
 * [ACRN Configuration](#acrn-configuration)
   - [ACRN BOARD Configuration](#acrn-board-configuration)
   - [ACRN SCENARIO Configruation](#acrn-scenario-configruation)
   - [ACRN BUILD MODE Configuration](#acrn-build-mode-configuration)
 * [Kernel Configuration](#kernel-configuration)
   - [SOS](#sos)
   - [UOS](#uos)
 * [GRUB Configuration](#grub-configuration)
 * [Override distro configuration](#override-distro-configuration)
 
VI.  [Build WIC Installer Image](#build-wic-installer-image)

VII. [Tested Hardware](#tested-hardware)

VIII.[Contributing](#contributing)

## Overview
This layer provides ACRN Hypvervisor integration with Yocto Project.

## Supported Hardware

ACRN is supported on below Intel platforms:
* Apollo Lake
* Kaby Lake
* Whiskey Lake
* Tiger Lake
* Elkhart Lake

> About minimum system requirements and limitations, please find more information at [Supported Hardware](https://projectacrn.github.io/latest/reference/hardware.html)

Author: This doc is for master-next branch, which is always on latest. So I think it is fine to point to latest here.
When I will backport to master, then I will do the necessary changes.

PREFERRED_PROVIDER_virtual/kernel = "linux-intel-acrn-sos"
```

#### Configure Post-launched User VM (UOS)

> Add below in `conf/multiconfig/uos.conf`:

Reviewer: "Add below" --> "Add the following"

This forwards ports 1-1 and 1-2 into the UOS, which on Skull Canyon is the two front ports (1-1 on the left, 1-2 on the right).  You'll need two input devices for this to work, obviously.

## Configurations

### MACHINE Configuration

> [meta-intel](https://git.yoctoproject.org/git/meta-intel) provides below machine configurations:

Reviewer: "provides below" --> "provides the following"

* intel-corei7-64

* intel-skylake-64

> intel-skylake-64 is 64-bit machine with -march=skylake and avx2 instruction-set set up. For more information, please check [intel-skylake-64.conf](http://git.yoctoproject.org/cgit/cgit.cgi/meta-intel/tree/conf/machine/intel-skylake-64.conf)

> intel-skylake-64 machine must be used for Skylake and successor platforms

> To configure MACHINE, set below in your `conf/local.conf`

Reviewer: "set below" --> "set the following"

```
MACHINE = "intel-corei7-64"
```

### Adding Guests

The multiconfig/package magic works with a `<image>-package.bb` recipe that inherits `container-package`. This puts the wic image and launcher script into a package which can be added as usual.

> To add core-image-base.wic image into Service OS (Add coe-image-base-package to your target ACRN image)

Reviewer 2: ```suggestion
To add core-image-base.wic image into Service OS (Add core-image-base-package to your target ACRN image)
```

- apl-up2-n3350
- apl-up2
- nuc6cayh
- nuc7i7dnb
- whl-ipc-i5
- whl-ipc-i7

For More information, Please check [Supported Hardware](https://projectacrn.github.io/latest/reference/hardware.html)

#### ACRN SCENARIO Configuration

To build for your acrn scenario, set `ACRN_SCENARIO` in your `conf/local.conf`. By default it is set to `industry` scenario.
```
ACRN_SCENARIO  = "hybrid"
```
Supported scenarios:
- sdc
- logical_partition
- industry
- hybrid
- hybrid_rt

For more information, please check [Build With the ACRN Scenario](https://projectacrn.github.io/latest/getting-started/building-from-source.html#build-with-the-acrn-scenario)

To customize ACRN Configruation, please check [Introduction to ACRN Configuration](https://projectacrn.github.io/latest/tutorials/acrn_configuration_tool.html)

#### ACRN BUILD MODE Configuration

To build ACRN in `RELEASE` mode, set `y` to `ACRN_RELEASE` in your `conf/local.conf`. By default it is set to `n`
```
ACRN_RELEASE = "y"
```

To build ACRN in `DEBUG` mode, set `n` to `ACRN_RELEASE` in your `conf/local.conf`.
```
ACRN_RELEASE = "n"
```

### Kernel Configuration

There are multiple kernel variant available for both SOS and UOS.

#### SOS

> To switch to linux-intel-acrn-sos LTS 5.4 kernel (default), in 'local.conf' replace with below lines:

Reviewer: "replace with below lines" --> "replace with the following lines" (also line 365 and additional occurrences further on)

To build for your acrn scenario, set `ACRN_SCENARIO` in your `conf/local.conf`. By default it is set to `industry` scenario.
```
ACRN_SCENARIO  = "hybrid"
```
Supported scenarios:
- sdc
- logical_partition
- industry
- hybrid
- hybrid_rt

For more information, please check [Build With the ACRN Scenario](https://projectacrn.github.io/latest/getting-started/building-from-source.html#build-with-the-acrn-scenario)

To customize ACRN Configruation, please check [Introduction to ACRN Configuration](https://projectacrn.github.io/latest/tutorials/acrn_configuration_tool.html)

#### ACRN BUILD MODE Configuration

To build ACRN in `RELEASE` mode, set `y` to `ACRN_RELEASE` in your `conf/local.conf`. By default it is set to `n`
```
ACRN_RELEASE = "y"
```

To build ACRN in `DEBUG` mode, set `n` to `ACRN_RELEASE` in your `conf/local.conf`.
```
ACRN_RELEASE = "n"
```

### Kernel Configuration

There are multiple kernel variant available for both SOS and UOS.

#### SOS

To switch to linux-intel-acrn-sos LTS 5.4 kernel (default), in 'local.conf' replace with below lines:
```
PREFERRED_PROVIDER_virtual/kernel = "linux-intel-acrn-sos"
PREFERRED_VERSION_linux-intel-acrn-sos = "5.4%"
```

> To switch to acrn-kernel-sos, in 'local.conf' replace with below lines:

Reviewer 2: What is the difference between this kernel and the `linux-intel-acrn-sos`? Perhaps add a note explaining this?

```
PREFERRED_PROVIDER_virtual/kernel = "acrn-kernel-sos"
```

#### UOS

To switch to linux-intel-acrn-uos LTS 5.4 kernel (default), in 'conf/multiconfig/uos.conf' replace with below lines:
```
PREFERRED_PROVIDER_virtual/kernel = "linux-intel-acrn-uos"
PREFERRED_VERSION_linux-intel-acrn-uos = "5.4%"
```

To switch to linux-intel-rt-acrn-uos Preempt-RT 5.4 kernel, in 'conf/multiconfig/uos.conf' replace with below line:
```
PREFERRED_PROVIDER_virtual/kernel = "linux-intel-rt-acrn-uos"
PREFERRED_VERSION_linux-intel-rt-acrn-uos = "5.4%"
```

To switch to acrn-kernel-uos, in 'conf/multiconfig/uos.conf' replace with below line:

```

> PREFERRED_PROVIDER_virtual/kernel = "acrn-kernel-uos"

Reviewer 2: Same question than for the sos kernel, what's the difference with the one previously mentionned?

```

### GRUB Configuration

> Below variables are used to prepare kernel command lines for Pre-launch VMs and Service OS. Based on these variables ACRN specific grub.cfg get generated.

Reviewer 2: ```suggestion
Below variables are used to prepare kernel command lines for Pre-launched User VMs and Service VM. Based on these variables an ACRN specific `grub.cfg` gets generated.
```

> These variables can be overwrite in your local config.

Reviewer 2: ```suggestion
These variables can be overwritten in your local config.
```

> * VMFLAGS - list of pre-launched VMs including Service vm

Reviewer 2: ```suggestion
* VMFLAGS - list of pre-launched VMs including Service VM
```

> * ACRN_EFI_GRUB2_MOD_CFG wic variable (semicolon (;) sperated list)

Reviewer 2: ```suggestion
* ACRN_EFI_GRUB2_MOD_CFG wic variable (semicolon (;) separated list)
```

    to make additional entries in grub.cfg i.e insmod ext3

> For More information, Please check [Update Ubuntu GRUB](https://projectacrn.github.io/latest/tutorials/using_hybrid_mode_on_nuc.html?highlight=hybrid%20mode#update-ubuntu-grub)

Reviewer 2: ```suggestion
For more information, please check [Update Ubuntu GRUB](https://projectacrn.github.io/latest/tutorials/using_hybrid_mode_on_nuc.html#update-ubuntu-grub)
```

and [acrn-bootconf.bbclass](https://github.com/intel/meta-acrn/blob/master-next/classes/acrn-bootconf.bbclass)

### Override distro configuration

> Due to parsing sequence conflict with meta-intel, weak assingments are not used in acrn-demo-sos and acrn-demo-uos distros. So to override distro configuration in local.conf, override syntex can be used i.e

Reviewer 2: ```suggestion
Due to parsing sequence conflict with `meta-intel`, weak assignments are not used in `acrn-demo-sos` and `acrn-demo-uos` distros. So to override distro configuration in `local.conf`, override syntax can be used i.e
```

```
WKS_FILE_acrn-demo-sos = "your-custom.wks.in"
```

## Build WIC Installer Image

To install the image on to NUC, you could burn the .wic image to the target hardware's internal storage.

Alternatively, you could build a wic based installer image where you can burn the .wic image onto USB flash drive and use USB flash drive as installer. To build the installer image for ACRN, add below lines to `conf/local.conf`:

```
BBMULTICONFIG_append  = " installer "
```

Add below in `conf/multiconfig/installer.conf`:

```
# use the installer wks file
WKS_FILE_pn-acrn-image-base = "image-installer.wks.in"

# build initramsfs to start the installation
INITRD_IMAGE_LIVE="core-image-minimal-initramfs"

# make sure initramfs and ext4 image are ready before building wic image
do_image_wic[depends] += "${INITRD_IMAGE_LIVE}:do_image_complete"
IMAGE_TYPEDEP_wic = "ext4"

> # content to be install

Reviewer 2: ```suggestion
# content to be installed
```

IMAGE_BOOT_FILES_append = "\
    ${KERNEL_IMAGETYPE} \
    acrn.bin;esp/acrn.bin \
    microcode.cpio;esp/microcode.cpio \
    grub-efi-bootx64.efi;EFI/BOOT/bootx64.efi \
    ${IMAGE_ROOTFS}/boot/EFI/BOOT/grub.cfg;esp/EFI/BOOT/grub.cfg \
    ${IMGDEPLOYDIR}/${IMAGE_BASENAME}-${MACHINE}.ext4;rootfs.img \
"
```

Now build the installer image:

```
$ bitbake mc:installer:acrn-image-base
```

> Currently ACRN WIC Installer is supported only OE-Core(poky) `master`, `hardknott` and `gatesgarth`. Support for `dunfell` is still in progress.

## Tested Hardware
The following undergo regular basic testing with their respective MACHINE types.

intel-corei7-64:
    NUC7i7DNH

intel-skylake-64:

>     NUC7i7DNH

Reviewer 2: So it's OK to use a SKL machine type for KBL?

Author: intel-skylake-64 machine should be use for Sky Lake and its successor platforms. So It is OK to use for KBL.
I have added comment in doc too.

## Contributing
You are encouraged to follow Github Pull request workflow to share changes and following commit message guidelines are recommended [OE patch guidelines](https://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines)

> Layer Maintainer: Naveen Saini

Reviewer 2: Perhaps add your email address (or Github profile)?
